File: classic/MergeSort.cpp
/*
 * a merge sort works as follows:
 * 1. Divide the unsorted list into n sublists, each containing 1 element is considered sorted
 * 2. Repeatedly merge sublists to produce new sorted sublists until there is only 1 sublist remaining. This will be the sorted list.
 * */
#include<iostream>
#include<vector>
#include<iterator>

using namespace std;

class Solution{
public:
	void mergesort(vector<int>& num,int l,int r){
		if(l<r){
			int m=l+(r-l)/2;
			mergesort(num,l,m);
			mergesort(num,m+1,r);
			merge(num,l,m,r);
			copy(num.begin(),num.end(),ostream_iterator<int>(cout,", "));
			cout<<endl;
		}
	}
	void merge(vector<int>& num,int l,int m,int r){
		int size1=m-l+1;
		int size2=r-m;
		vector<int> left(num.begin()+l,num.begin()+m+1);	
		vector<int> right(num.begin()+m+1,num.begin()+r+1);
		int i=0,j=0,k=l;
		while(i<size1&&j<size2){
			if(left[i]<=right[j]){
				num[k++]=left[i++];
			}
			else{
				num[k++]=right[j++];
			}
		}
		while(i<size1)	{
			num[k++]=left[i++];
		}
		while(j<size2){
			num[k++]=right[j++];
		}

	}
};


int main(){
	Solution s;
	vector<int> num={2,1,12,11,19,17,8,13,5,6,7};
	s.mergesort(num,0,num.size()-1);
	
	copy(num.begin(),num.end(),ostream_iterator<int>(cout,", "));
	cout<<endl;
	return 0;
}

File: classic/MergeSort.py
def mergeSort(s, l, r):
    if l>=r:
        return
    mid = (l+r)/2
    mergeSort(s, l, mid)
    mergeSort(s, mid+1, r)
    merge(s, l, mid, r)

def merge(s, l, mid, r):
    t = []
    i = l
    j = mid+1
    while i<=mid and j<=r:
        if s[i] < s[j]:
            t.append(s[i])
            i+=1
        else:
            t.append(s[j])
            j+=1
    while i<=mid:
        t.append(s[i])
        i+=1
    while j<=r:
        t.append(s[j])
        j+=1
    for i, x in enumerate(t):
        s[l+i] = x

if __name__=="__main__":
    s = [3,5,1,2,7,8,10,0]
    mergeSort(s, 0, len(s)-1)
    print s

    s = [9,8,7,6,5,4,3,2,1]
    mergeSort(s, 0, len(s)-1)
    print s
